cyrusimap/cyrus-imapd
# Cassandane::Tiny - cassandane tests in many small files

🧪 This is sort of a proof of concept, but possibly good enough to move forward.  (I do think we can improve on this quite a bit, though.)

## Why?

The chief problem I'm trying to solve is merge conflicts.  Right now, we have a few *very* large test files (like JMAPEmail) to which tests get appended.  When there are several active branches, this can result in a bunch of conflict with the tests all being added in one place.  To avoid that, we make every test a single file.  Then the main conflict is that we must not use the same name more than once.

## Cassandane::Tiny

The mechanism here is pretty gross, but I've tried to make it sort of safe for future fix-ups.

Cassandane tests use Test::Unit, in which subroutines with names beginning with `test_` are tests.  The test runner loads the classes, gets a list of their methods, and runs them.  This branch introduces **Cassandane::Tiny::Loader**, which is invoked with a path to a tree of test files.  Each test file is loaded into a test class, and is expected to contain one test subroutine.  Nothing enforces this, but every test must indicate that it's meant to be loaded this way by including `use Cassandane::Tiny`.  Later, we may provide an alternate boilerplate module.

I split JMAPCalendar and JMAPEmail to use tiny tests, found in ./cassandane/tiny-tests subdirectories.

## Selective Running

It's already possible to run `testrunner JMAPEmail/email_set_foo`.  This branch adds, as somewhat unrelated work, the ability to run `testrunner JMAPEmail/email_set_*`, which will match anything starting with `email_set`.  You can also use negated tests, so that…

```
$ ./testrunner JMAPEmail '!JMAPEmail/email*'
```

…means "run all the JMAPEmail tests that don't have a name starting with "email".

## Next Steps

I'd like to be able to ditch Test::Unit over time, but this seems like a very big project.

We should convert more tests into tiny tests.  (Process described below.)

We should eliminate the difference between any two classes loading tiny tests.  That is: test selection by wildcard means that there's not much reason for individual test classes.  We should be able to make a single class that loads all tests.  This means figuring out what methods in each class can be made common, and how to resolve conflicts.  *We can do this over a long time if we like.*

I'd like to ensure that the filename and test name are directly related, rather than being similar by convention.  This can be done later.

## How to convert a Cassandane test class to tiny tests

This could be more automated, but I'm not sure it'd be worth the time…

1. reorganize the `.pm` file so that all the `test_` subroutines are at the end and all the non-test subroutines are ahead
2. write all the test subroutines to a new file (called, say "foo")
3. delete them from the `.pm` file — make sure you leave the `1;` in the pm file
4. in place of the test subroutines, add `use Cassandane::Tiny::Loader 'foo';`
5. run `perl tiny-test-splitter.pl foo foo`
6. add `tiny-tests/foo` and commit everything
7. make sure it works!

My most common mistake was getting warnings out of step 5 and realizing that I'd missed some subs in 2/3.
